Give more explanatory error for bad 'Mesh/block' specification

lindsayad opened the pull request:
Closes #22928

lindsayad committed:
Give more explanatory error for bad 'Mesh/block' specification

Closes #22928

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1283937/) on 2e38e75 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23077/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Test install](https://civet.inl.gov/job/1287922/) on dbac307 : invalidated by @lindsayad

lindsayad committed:
Make code honor code comments

If the coordinate system type parameter is size 0 or size 1, then
we assume the user wants to set the coordinate types for all subdomains,
not just the ones specified by a block or coord_block parameter. This
is consistent with the code comments

lindsayad commented:
Ok @dschwen I was able to make your tutorial inputs work without requiring any modification

lindsayad committed:
Change Mesh/block to Mesh/coord_block

moosebuild commented:
Job [Nonsparse AD](https://civet.inl.gov/job/1288107/) on a9b05a8 : invalidated by @lindsayad

moosebuild commented:
Job [LibTorch](https://civet.inl.gov/job/1288103/) on a9b05a8 : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1288124/) on a9b05a8 wanted to post the following:


## Framework coverage

Mutiple base shas were found when retrieving head reports:
  fe98211c5cb7fddf763b8db977bc5a558912402e
  c167ac36eb4dfad1ded966dd321f719a3763e96c

## Modules coverage

Mutiple base shas were found when retrieving head reports:
  c167ac36eb4dfad1ded966dd321f719a3763e96c
  fe98211c5cb7fddf763b8db977bc5a558912402e

## Full coverage reports

<details>
<summary>Reports</summary>

- `framework` (FAILED)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/solid_properties)
- `stochastic_tools` (FAILED)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23077/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1288124/) on a9b05a8 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)

lindsayad committed:
Don't do conversion to set and back

moosebuild commented:
Job [Python 3.9](https://civet.inl.gov/job/1290281/) on 4564661 : invalidated by @lindsayad

lindsayad committed:
Protect against mesh not being prepared in setCoordSystem

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/1291128/) on b29991b : invalidated by @lindsayad

lindsayad commented:
@dschwen would you be willing to review this?

dschwen commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -177,7 +177,7 @@ FEProblemBase::validParams()
   /// 3. _blocks.size() > 0 and one coordinate system was specified, then the whole domain will be that system.
   params.addDeprecatedParam<std::vector<SubdomainName>>(
-      "block", "Block IDs for the coordinate systems", "Please use 'Mesh/block' instead");
+      "block", "Block IDs for the coordinate systems", "Please use 'Mesh/coord_block' instead");
```
Are GlobalParams still applied to deprecated params? If so then this PR will not directly fix my issues, but once the deprecated parameter is removed it should

lindsayad commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -3523,72 +3535,62 @@ MooseMesh::setCoordSystem(const std::vector<SubdomainName> & blocks,
+    const auto coord_type = coord_sys.size() == 0
+                                ? Moose::COORD_XYZ
+                                : Moose::stringToEnum<Moose::CoordinateSystemType>(coord_sys[0]);
+    for (const auto sid : subdomains)
```
@dschwen this fixes your issue regardless of the fact that `block` gets picked up from `GlobalParams`. As long as `coord_sys` is size 1 (or size 0) then we have changed to automatically set *all* mesh subdomains to that coordinate system type, regardless of the value of `block`. Before we just set it for the specifically set blocks
